rir/src/compiler/translations/rir_2_pir/rir_2_pir_compiler.h

```diff
@@ -11,17 +13,22 @@ class Rir2PirCompiler : public RirCompiler {
   public:
     Rir2PirCompiler(Module* module, const DebugOptions& debug);
 
-    void compileClosure(SEXP, MaybeCls success, Maybe fail) override;
+    void compileClosure(SEXP, MaybeCls success, Maybe fail,
+                        bool isIndependent) override;
     void compileFunction(rir::Function*, FormalArgs const&, MaybeCls success,
                          Maybe fail);
     void optimizeModule();
     void printAfterPass(const std::string&, const std::string&, Closure*,
                         size_t);
 
+    StreamLogger& getLog() { return log; }
```

Comment:
Making it a constant is a mess because then the compiler requires to label as constant all the public functions.
I do not get why it hurts to access the member through a getter. Is it a c++ convention not to do that?

Reply:
I agree that the consts can sometimes be a bit like cancer, you need to put them everywhere. So I think it is fine to have the StreamLogger be mutuable by any class that can touch it, that seems fine. But I think the issue here is, why is `getLog()` public? it seems to me that the logger should only be used internally and there is no need for a *client* of the rir_2_pir_compiler to access the logger.

Reply:
as to the question, is a getter not idiomatic: At least for may personal taste a getter, that has no logic and is only used internally is very strange.

Usually my preference is:

1. private member, no getter
2. external client needs to access: private member, public getter

Const or non-const is an orthogonal issue. It's kind of hard to grasp at the beginning.

`const StreamLogger& getLog()` means that it will return you a const reference. That means only methods marked const can be executed on the reference that you get from `getLog`. Usually if I have a member that i can initialize in the constructor and never need to modify during its lifetime, I make it const.

If I have to modify it internally, but no external client should ever modify it, then I try to return a const reference/ptr. But as you say, in particular if you have const and non-const refs it might be annoying.

The advantage of a const member is, that the list above changes to:

1. private member, no getter
2. only read access to member by client : public const member, no getter
3. write access to member by client : private member, public non-const getter

If the member is const, then it is fine to make it public, since the client should not break your invariants, since the internal object you give it const. Thus you can avoid having a getter.

Reply:
> At least for my personal taste a getter, that has no logic and is only used internally is very strange.

I basically agree, that kind of getter seems a bit pointless. If it's used internally, then there's no need for a public interface. If there's no logic and it just returns a member, then why not just access the member directly?

If the getter is used externally, it makes sense to hide the implementation details. Even if there's no logic, maybe one day we want to change the implementation and add some logic without breaking clients. But in this case, for something like the logger, changes to the implementation will probably break the client anyway, so a getter wouldn't really hide anything.

Reply:
So, in this case the member is used by clients. The `StreamLogger` is owned by `Rir2PirCompiler`and used elsewhere by the compilation phases. To avoid passing it as parameter I added the getter so that any client already having a reference to the compiler can access the log. The situation is quite similar in the `Pir2RirCompiler`. The class actually doing stuff and needing the logger is `Pir2Rir`which has a reference to the compiler. So, at least under this model I do not see an alternative. 

Then, the getter can not be *const* because the clients use startLogging which uses emplace that need a *non-const* member. And finally, I could remove the reference and make the member just public ok. But I still do not see how the getter hurts. And from why I understand from Oli's quote, he is not againts it: 
> 3. write access to member by client : private member, public non-const getter

Reply:
I don't think it hurts, but I wanted to clarify why I didn't really like it in the first place. Maybe the problem is that `Rir2PirCompiler` owns the `StreamLogger` (and `Pir2RirCompiler` has a reference)?

Anyway, I won't push any more on this, it's already been enough of a detour!